File: Applications/Python/Stack/GenerateParantheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack = []
        result = []

        def rec(o, c):
            if o == n and c == n:                
                result.append("".join(stack))
                return 
            
            if o < n:
                stack.append("(")
                rec(o + 1, c)
                stack.pop()
            
            if c < o:
                stack.append(")")
                rec(o, c + 1)
                stack.pop()

        rec(0, 0)

        return result

File: Applications/CSharp/Stack/GenerateParantheses.cs
class Solution {
    public IList<string> GenerateParenthesis(int n) {
        var result = new List<string>();
        var stack = new Stack<char>();

        void Rec(int c, int o)
        {
            if(c == n && o == n)
            {                
                 result.Add(new string(stack.Reverse().ToArray()));
            }

            if (c < n)
            {
                stack.Push('(');
                Rec(c + 1, o);
                stack.Pop();
            }

            if(o < c)
            {
                stack.Push(')');
                Rec(c, o + 1);
                stack.Pop();
            }
        }

        Rec(0, 0);
        return result;   
    }
}